File: src/lib/pricing/pool-risk.ts
/**
 * Риск-движок пула: неттинг лонг/шорт и лимиты на приём позиций.
 * Порт services/leverage/pool_risk.py.
 *
 * Лимиты — в терминах МАКСИМАЛЬНОЙ ВЫПЛАТЫ (обязательства пула), а не
 * номинала. Три уровня: на рынок, на кластер коррелированных рынков, на весь
 * пул. Off-chain-логика «предлагает»; в контракте эти лимиты станут
 * инвариантами.
 */

import { LONG, type Position } from "./knockout";
import { roundHalfEven } from "./ticks";

export interface PoolLimits {
  /** Весь капитал пула, $. */
  capital: number;
  /** ≤3% капитала на один рынок. */
  perMarket: number;
  /** ≤10% на кластер коррелированных рынков. */
  perCluster: number;
  /** ≤30% суммарно по всему пулу. */
  globalStress: number;
}

export const DEFAULT_POOL_SHARES = {
  perMarket: 0.03,
  perCluster: 0.1,
  globalStress: 0.3,
} as const;

/** Лимиты с их долями по умолчанию. */
export function poolLimits(capital: number, overrides?: Partial<PoolLimits>): PoolLimits {
  return { capital, ...DEFAULT_POOL_SHARES, ...overrides };
}

/** Обязательства пула по одному рынку, разбитые на лонг/шорт. */
export class Exposure {
  /** Сумма maxPayout лонгов. */
  longPayout: number;
  /** Сумма maxPayout шортов. */
  shortPayout: number;

  constructor(longPayout = 0, shortPayout = 0) {
    this.longPayout = longPayout;
    this.shortPayout = shortPayout;
  }

  /** Чистый перекос = |лонги − шорты|. Это и есть риск пула по рынку. */
  get net(): number {
    return Math.abs(this.longPayout - this.shortPayout);
  }
}

/** Вердикт по позиции: можно ли принять и почему нет. */
export interface PoolVerdict {
  ok: boolean;
  reason: string;
}

export class PoolRiskEngine {
  readonly limits: PoolLimits;
  /** marketId → clusterId. Нет записи — рынок сам себе кластер. */
  readonly clusters: Record<string, string>;
  readonly exposure: Map<string, Exposure>;

  constructor(limits: PoolLimits, clusters?: Record<string, string>) {
    this.limits = limits;
    this.clusters = clusters ?? {};
    this.exposure = new Map();
  }

  private clusterOf(marketId: string): string {
    return this.clusters[marketId] ?? marketId;
  }

  /** Копия экспозиции рынка с добавленной позицией — для проверки ДО принятия. */
  private withAdded(pos: Position): Exposure {
    const cur = this.exposure.get(pos.marketId) ?? new Exposure();
    const next = new Exposure(cur.longPayout, cur.shortPayout);
    if (pos.direction === LONG) next.longPayout += pos.maxPayout;
    else next.shortPayout += pos.maxPayout;
    return next;
  }

  /** Таблица экспозиций с наложенным пробным изменением. */
  private table(override?: Map<string, Exposure>): Map<string, Exposure> {
    if (!override) return this.exposure;
    const merged = new Map(this.exposure);
    for (const [id, exp] of override) merged.set(id, exp);
    return merged;
  }

  /** Суммарный чистый риск по кластеру (сумма перекосов его рынков). */
  netCluster(clusterId: string, override?: Map<string, Exposure>): number {
    let total = 0;
    for (const [marketId, exp] of this.table(override)) {
      if (this.clusterOf(marketId) === clusterId) total += exp.net;
    }
    return total;
  }

  /** Суммарный чистый риск по всему пулу. */
  netGlobal(override?: Map<string, Exposure>): number {
    let total = 0;
    for (const exp of this.table(override).values()) total += exp.net;
    return total;
  }

  /**
   * Проверить, можно ли принять позицию, НЕ меняя состояние.
   * Проверяет все три лимита и возвращает причину отказа.
   */
  canAccept(pos: Position): PoolVerdict {
    const trial = this.withAdded(pos);
    const override = new Map([[pos.marketId, trial]]);
    const cap = this.limits.capital;

    // 1) лимит на рынок
    const marketCap = cap * this.limits.perMarket;
    if (trial.net > marketCap) {
      return {
        ok: false,
        reason: `per_market: перекос ${roundHalfEven(trial.net)} > лимит ${roundHalfEven(marketCap)}`,
      };
    }

    // 2) лимит на кластер
    const cluster = this.clusterOf(pos.marketId);
    if (this.netCluster(cluster, override) > cap * this.limits.perCluster) {
      return { ok: false, reason: "per_cluster: превышен лимит кластера" };
    }

    // 3) глобальный лимит
    if (this.netGlobal(override) > cap * this.limits.globalStress) {
      return { ok: false, reason: "global_stress: превышен глобальный лимит" };
    }

    return { ok: true, reason: "ok" };
  }

  /** Принять позицию — обновить экспозицию рынка. Зовётся ПОСЛЕ canAccept. */
  accept(pos: Position): void {
    const verdict = this.canAccept(pos);
    if (!verdict.ok) throw new Error(`позиция отклонена: ${verdict.reason}`);

    const cur = this.exposure.get(pos.marketId) ?? new Exposure();
    if (pos.direction === LONG) cur.longPayout += pos.maxPayout;
    else cur.shortPayout += pos.maxPayout;
    this.exposure.set(pos.marketId, cur);
  }
}

File: backend/services/leverage/pool_risk.py
# services/leverage/pool_risk.py
"""
Риск-движок пула: неттинг лонг/шорт и лимиты на приём позиций.

Лимиты — в терминах МАКСИМАЛЬНОЙ ВЫПЛАТЫ (обязательства пула), не номинала.
Три уровня: на рынок, на кластер коррелированных рынков, на весь пул.

Off-chain-логика "предлагает"; в контракте эти лимиты станут инвариантами.
"""
from __future__ import annotations

from dataclasses import dataclass, field

from knockout import Position, LONG, SHORT


@dataclass(frozen=True)
class PoolLimits:
    capital: float               # весь капитал пула, $
    per_market: float = 0.03     # ≤3% капитала на один рынок
    per_cluster: float = 0.10    # ≤10% на кластер коррелированных рынков
    global_stress: float = 0.30  # ≤30% суммарно по всему пулу


@dataclass
class Exposure:
    """Обязательства пула по одному рынку, разбитые на лонг/шорт."""
    long_payout: float = 0.0   # сумма max_payout лонгов
    short_payout: float = 0.0  # сумма max_payout шортов

    @property
    def net(self) -> float:
        """Чистый перекос = |лонги − шорты|. Это и есть риск пула по рынку."""
        return abs(self.long_payout - self.short_payout)


class PoolRiskEngine:
    def __init__(self, limits: PoolLimits, clusters: dict[str, str] | None = None):
        self.limits = limits
        # market_id -> cluster_id (какие рынки коррелируют). Нет записи -> сам себе кластер.
        self.clusters = clusters or {}
        self.exposure: dict[str, Exposure] = {}

    def _cluster_of(self, market_id: str) -> str:
        return self.clusters.get(market_id, market_id)

    def _with_added(self, pos: Position) -> Exposure:
        """Копия экспозиции рынка с добавленной позицией — для проверки ДО принятия."""
        cur = self.exposure.get(pos.market_id, Exposure())
        new = Exposure(cur.long_payout, cur.short_payout)
        if pos.direction == LONG:
            new.long_payout += pos.max_payout
        else:
            new.short_payout += pos.max_payout
        return new

    def net_cluster(self, cluster_id: str, override: dict[str, Exposure] | None = None) -> float:
        """Суммарный чистый риск по кластеру (сумма перекосов его рынков)."""
        table = {**self.exposure, **(override or {})}
        total = 0.0
        for mid, exp in table.items():
            if self._cluster_of(mid) == cluster_id:
                total += exp.net
        return total

    def net_global(self, override: dict[str, Exposure] | None = None) -> float:
        """Суммарный чистый риск по всему пулу."""
        table = {**self.exposure, **(override or {})}
        return sum(exp.net for exp in table.values())

    def can_accept(self, pos: Position) -> tuple[bool, str]:
        """
        Проверить, можно ли принять позицию, НЕ меняя состояние.
        Возвращает (можно?, причина_отказа). Проверяет все три лимита.
        """
        trial = self._with_added(pos)
        override = {pos.market_id: trial}

        # 1) лимит на рынок
        cap = self.limits.capital
        if trial.net > cap * self.limits.per_market:
            return False, f"per_market: перекос {trial.net:.0f} > лимит {cap*self.limits.per_market:.0f}"

        # 2) лимит на кластер
        cluster = self._cluster_of(pos.market_id)
        if self.net_cluster(cluster, override) > cap * self.limits.per_cluster:
            return False, "per_cluster: превышен лимит кластера"

        # 3) глобальный лимит
        if self.net_global(override) > cap * self.limits.global_stress:
            return False, "global_stress: превышен глобальный лимит"

        return True, "ok"

    def accept(self, pos: Position) -> None:
        """Принять позицию — обновить экспозицию рынка. Зовётся ПОСЛЕ can_accept."""
        ok, reason = self.can_accept(pos)
        if not ok:
            raise ValueError(f"позиция отклонена: {reason}")
        cur = self.exposure.get(pos.market_id, Exposure())
        if pos.direction == LONG:
            cur.long_payout += pos.max_payout
        else:
            cur.short_payout += pos.max_payout
        self.exposure[pos.market_id] = cur